controller/lib/core/flow_integrator.cpp

```diff
@@ -14,63 +14,78 @@ limitations under the License.
-// we'd need an error of +/-20ml in each.  At zero flow, this corresponds to
-// significantly less than 1 Pa of pressure, and just two or three bits in our
-// 12-bit ADC!  See
-// https://docs.google.com/spreadsheets/d/1G9Kb-ImlluK8MOx-ce2rlHUBnTOtAFQvKjjs1bEhlpM
-//
-// In theory the story is better at higher flows (because the venturi response
-// is the square root of pressure), and we may run a bias flow through our
-// device to put us into a regime where we're less sensitive to error
-// (equivalently, where the venturi response has a lower slope).  But we still
-// want to handle the low-flow case as best we can.
-static constexpr float FLOW_PID_KU = 0.20f;
-static constexpr float FLOW_PID_TU = 5;
-static constexpr VolumetricFlow FLOW_PID_OUTPUT_MAX = ml_per_sec(100);
-
-// TODO: Make PID parameters configurable.
-
-FlowIntegrator::FlowIntegrator()
-    : flow_correction_pid_(
-          /*kp=*/0.6f * FLOW_PID_KU,
-          /*ki=*/1.2f * FLOW_PID_KU / FLOW_PID_TU,
-          /*kd=*/3 * FLOW_PID_KU * FLOW_PID_TU / 40,
-          // ProportionalTerm::ON_ERROR and DifferentialTerm::ON_MEASUREMENT
-          // makes for your "standard" PID.  TODO: perhaps these shouldn't even
-          // be options.
-          ProportionalTerm::ON_ERROR,        //
-          DifferentialTerm::ON_MEASUREMENT,  //
-          -FLOW_PID_OUTPUT_MAX.ml_per_sec(), //
-          FLOW_PID_OUTPUT_MAX.ml_per_sec()) {}
+FlowIntegrator::FlowIntegrator() = default;
 
 void FlowIntegrator::AddFlow(Time now, VolumetricFlow uncorrected_flow) {
-  VolumetricFlow flow = uncorrected_flow + flow_correction_;
-  Duration delta = now - last_flow_measurement_time_;
-  if (delta >= VOLUME_INTEGRAL_INTERVAL) {
-    volume_ += delta * (last_flow_ + flow) / 2.0f;
+  VolumetricFlow flow = uncorrected_flow + FlowCorrection();
+
+  // First time?  In that case, don't need to adjust the flow.  Just record
+  // this data point.
+  if (last_flow_measurement_time_ == std::nullopt) {
+    last_note_time_ = now;
+    last_note_vol_ = ml(0);
+    last_flow_measurement_time_ = now;
+    last_flow_ = flow;
+    return;
+  }
+
+  Duration dt = now - *last_flow_measurement_time_;
+  if (dt >= VOLUME_INTEGRAL_INTERVAL) {
+    volume_ += dt * (last_flow_ + flow) / 2.0f;
     last_flow_measurement_time_ = now;
     last_flow_ = flow;
   }
 }
 
 void FlowIntegrator::NoteExpectedVolume(Volume v) {
-  flow_correction_ = ml_per_sec(flow_correction_pid_.Compute(
-      last_flow_measurement_time_, volume_.ml(), v.ml()));
+  // We need at least one call to AddFlow before we can do anything.
+  if (last_note_time_ == std::nullopt ||
+      last_flow_measurement_time_ == std::nullopt ||
+      *last_note_time_ >= *last_flow_measurement_time_) {
+    return;
+  }
+
+  // Compute the new flow correction as follows.
+  //
+  // - Compute the time delta between now (last_flow_measurement_time_) and the
+  //   last call to NoteExpectedVolume.
+  Duration dt = *last_flow_measurement_time_ - *last_note_time_;
```

Comment:
Rename to dt_since_last_note_? In code below I confused it for "dt between integration steps".